### ISSUE TITLE
Validate dns challenge via dedicated second domain

### ISSUE THREAD

**Issue by captncraig** [xenolf/lego#479]
I have an interesting use case that may or may not make sense to integrate into this project. Consider the following situation:

- I have many domains that I want on a single SAN cert.
- They have dns hosted in different places (and some in multiple places)
- Organization may be reluctant to provide dns credentials to some or all of those domains for LE purposes

 I had an idea though when I saw someone mention that the Let's Encrypt validator follows CNAME records:

1. Set up permanent `CNAME _acme-challenge.example.com` to point at `_acme-challenge.example.com.otherdomain.tld`
2. To validate the domain, you just need to set the appropriate TXT record on `otherdomain.tld`, and LE can validate that.

This saves you from having to know how to update DNS on each individual domain, as long as you can edit the dedicated matching domain.

I guess the first question is, "Is this general purpose enough to consider for inclusion in this repo"? 

My [proof of concept](https://github.com/xenolf/lego/compare/master...captncraig:transform?expand=1) works as another provider that acts as a wrapper around the real provider you are using. It is fairly limited in how it can transform the record, because of the current way providers generate their own fqdns. But it does work. 

Positives:

- Lets you issue certs on domains that you cannot directly control, as long as you have previously created the appropriate CNAME. 
- Works fairly seamlessly with other providers, and cli (currently just set `--dns transform` along with `TRANSFORM_PROVIDER` and `TRANSFORM_DOMAIN` env vars)

Negative:

- kinda tricky to understand what exactly is going on. Might be kinda niche.
- subdomain certs require a longer, more verbose CNAME setup. I wish a wildcard CNAME could  cover it, but that isn't feasable. In the future I think wildcard certs should only need one CNAME though per wildcard pattern.
- current code path where each provider calls acme.DNS01Record to get the fqdn prevents me from doing other patterns, like have all the cnames point to a single or simpler name on the validation domain. But whatever.

Any thoughts? Should I submit a PR from what I have?

**Comment by captncraig** [comment]
After a train ride to think about this, I think I've talked myself out of it. It feels like a bit too much yak shaving. I can't explain the process cleanly to my team without a whole lot of "you want me to do what?". 

So disregard this. I'm probably going to levera our [existing dns management platform](https://github.com/StackExchange/dnscontrol) as a challenge provider shim, since that already has a clear view of the layout of our domains and providers. I'll have a simple command that gathers the desired certs and uses lego as a library to issue certs after we initialize the provider shims appropriately on a per-domain basis.

**Comment by captncraig** [comment]
If anyone thinks this might be a better idea than I do, feel free to look at the branch I linked in the original post and reconsider some variation of it.

**Comment by mholt** [comment]
Thanks Craig, for the detailed proposal and PoC! I agree, it's quite niche: however, @xenolf might think otherwise (I dunno). At least if anyone else has this need, they can comment here and we can get a gauge for how much demand there is.

**Comment by mhoran** [comment]
I *think* this is something that I want. My case is slightly simpler.

I would like to get a certificate for `a.example.com`. `example.com` DNS is hosted on a DNS server running nsd. I can of course delegate any subdomain to another name server, e.g. route53. Therefore, I've delegated `a.acme.example.com` to route53. I then have set up `_acme-challenge.a.example.com CNAME acme-challenge.a.acme.example.com`. I would like for lego to support setting the target of the CNAME, but it can't do that today without modification.

I've hacked the route53 provider to do what I want by providing yet another environment variable. I can also see a way to get what I want from lego by modifying `FindZoneByFqdnCustom` to traverse CNAMEs. I'm wondering whether either of these would be welcome changes, or if, like @captncraig, this use case is too niche. If so, I guess I can write an exec provider.

Note that I got the idea for the CNAME solution from this [1] blog post from EFF. So perhaps this configuration is not so far fetched?

[1] https://www.eff.org/deeplinks/2018/02/technical-deep-dive-securing-automation-acme-dns-challenge-validation
